### PR TITLE
Using binary search in getCandidateStickyLinesIntersectingFromOutline in order to improve performance

### LINKED ISSUE #156874: Sticky Scroll is slow for large files
* it would be good if rendering the sticky scroll would not iterate over all the ranges
* you can try with https://github.com/microsoft/TypeScript/blob/main/src/compiler/checker.ts
* idea we discussed:
  * when creating the ranges, capture the parent index for each one of them
  * when rendering, do a binary search for the first line(s) in the viewport

### PR DESCRIPTION
**Pull request by aiday-mar**
Fixes #156874

### CODE REVIEW

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 1]
```diff
@@ -96,10 +96,44 @@ export class StickyLineCandidateProvider extends Disposable {
 		}
 	}
 
+	private customBinarySearch(model: readonly StickyOutlineElement[], startLine: number) {
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:99]: There is [binarySearch](https://github.com/microsoft/vscode/blob/e7f76d71041ddb7e73bdf0cf258b811902dc4c8c/src/vs/base/common/arrays.ts#L72) and [binarySearch2](https://github.com/microsoft/vscode/blob/e7f76d71041ddb7e73bdf0cf258b811902dc4c8c/src/vs/base/common/arrays.ts#L91) and you have a very strong for not using any of those ;-)

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 2]
```diff
@@ -96,10 +96,44 @@ export class StickyLineCandidateProvider extends Disposable {
...
+		if (result === -1) {
+			result = 0;
+		} else if (result < 0) {
+			result = -result - 2;
+		}
+		return result;
+	}
+
 	public getCandidateStickyLinesIntersectingFromOutline(range: StickyRange, outlineModel: StickyOutlineElement, result: StickyLineCandidate[], depth: number, lastStartLineNumber: number): void {
+		if (!outlineModel || outlineModel.children.length === 0) {
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:128]: The `!outlineModel` condition should never be true or the argument typing is wrong, should be `outlineModel: StickyOutlineElement | undefined`

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 3]
```diff
@@ -133,9 +167,13 @@ export class StickyLineCandidateProvider extends Disposable {
...
-		const children = [...outlineModel.children.values()].map(child =>
-			StickyOutlineElement.fromOutlineModel(child)
-		);
+		const children = [...outlineModel.children.values()].map(child => {
+			if (child instanceof OutlineElement && child.symbol.selectionRange.startLineNumber !== child.symbol.range.endLineNumber || child instanceof OutlineGroup || child instanceof OutlineModel) {
+				return StickyOutlineElement.fromOutlineModel(child);
+			} else {
+				return;
+			}
+		}).filter((child) => !!child) as StickyOutlineElement[];
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:176]: Looking at this with the performance-eye: 

* the `[...something.values()]`-statement iterates over all values and allocates a new array
* the `map`-call iterates again over all values and also creates a new array
* the `filter`-call iterates again and creates yet another array

It's better to iterate just once using a forof-loop and some if-statement

Comment (aiday-mar) [reply src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:176]: That's right, thanks for pointing it out

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 4]
```diff
@@ -96,9 +97,28 @@ export class StickyLineCandidateProvider extends Disposable {
...
+		return index;
+	}
+
 	public getCandidateStickyLinesIntersectingFromOutline(range: StickyRange, outlineModel: StickyOutlineElement, result: StickyLineCandidate[], depth: number, lastStartLineNumber: number): void {
+		if (outlineModel.children.length === 0) {
+			return;
+		}
 		let lastLine = lastStartLineNumber;
-		for (const child of outlineModel.children) {
+		const childrenStartLines = outlineModel.children.map(child => child.range?.startLineNumber as number);
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:114]: The `childrenStartLines`-array can contain `undefined` and numbers. That is likely unexpected and maybe something that you should check for. Like, ignore those element that have no range

**Review by jrieken — APPROVED** [review]

**Pull request merged**